src/java/main/org/apache/zookeeper/server/OldEphemeralType.java

```diff
@@ -0,0 +1,74 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server;
+
+/**
+ * See https://issues.apache.org/jira/browse/ZOOKEEPER-2901
+ *
+ * version 3.5.3 introduced bugs associated with how TTL nodes were implemented. version 3.5.4
+ * fixes the problems but makes TTL nodes created in 3.5.3 invalid. OldEphemeralType is a copy
+ * of the old - bad - implementation that is provided as a workaround. {@link EphemeralType#TTL_3_5_3_EMULATION_PROPERTY}
+ * can be used to emulate support of the badly specified TTL nodes.
+ */
+public enum OldEphemeralType {
+    /**
+     * Not ephemeral
+     */
+    VOID,
+    /**
+     * Standard, pre-3.5.x EPHEMERAL
+     */
+    NORMAL,
+    /**
+     * Container node
+     */
+    CONTAINER,
+    /**
+     * TTL node
+     */
+    TTL;
+
+    public static final long CONTAINER_EPHEMERAL_OWNER = Long.MIN_VALUE;
+    public static final long MAX_TTL = 0x0fffffffffffffffL;
+    public static final long TTL_MASK = 0x8000000000000000L;
+
+    public static OldEphemeralType get(long ephemeralOwner) {
```

Comment:
Do think it has value to keep this "old" type in the codebase?
For backward compatibility reason I don't think it makes sense on a non-stable branch.
Its contents only used from the "new" type and tests.
I'd rather put the logic which is still needed in the "new" type and get rid of this one completely keeping EphemeralType nice and clean.
What do you think?

Reply:
We should create a new task to delete it after a while. However, I know that this will be important. Whoever used TTLs in 3.5.3 will run into problems when they upgrade. We need to have a workaround for these users.

Reply:
Yes, that's fine with the emulate353 flag.
My concern is that we keep 2 enums in the codebase: `EphemeralType` and `OldEphemeralType` while I think it'd be nicer to build the functionality of the old enum into the new one.

Reply:
I put it in `OldEphemeralType` so it's easier to remove and reason about. The emulation is a separate concern.

Reply:
Makes sense.
I think it would be slightly more accurate to name the old enum to `EphemeralTypeEmu353`.
What do you think?